clientv3/balancer.go

```diff
@@ -162,6 +163,18 @@ func (b *simpleBalancer) Up(addr grpc.Address) func(error) {
 
 func (b *simpleBalancer) Get(ctx context.Context, opts grpc.BalancerGetOptions) (grpc.Address, func(), error) {
 	var addr string
+
```

Comment:
With your idea, I got this:

```go
func (b *simpleBalancer) Get(ctx context.Context, opts grpc.BalancerGetOptions) (grpc.Address, func(), error) {
        var addr string

        // If opts.BlockingWait is false (for fail-fast RPCs), it should return
        // an address it has notified via Notify immediately instead of blocking.
        blockC := make(chan struct{})
        if !opts.BlockingWait {
                close(blockC)
        }

        for {
                b.mu.RLock()
                ch := b.upc
                b.mu.RUnlock()
                select {
                case <-ch:
                case <-blockC:
                case <-ctx.Done():
                        return grpc.Address{Addr: ""}, nil, ctx.Err()
                }
                b.mu.RLock()
                addr = b.pinAddr
                upEps := len(b.upEps)
                b.mu.RUnlock()
                if !opts.BlockingWait {
                        if upEps == 0 {
                                return grpc.Address{Addr: ""}, nil, NoAddressAvailable
                        }
                        break
                }
                if addr == "" {
                        return grpc.Address{Addr: ""}, nil, grpc.ErrClientConnClosing
                }
                if upEps > 0 {
                        break
                }
        }
        return grpc.Address{Addr: addr}, func() {}, nil
}
```

If we merge `ctx.Done()` to `waitC`, returning `ctx.Err()` in the `select` will be hard in the following part. I therefore defined a separate channel.
However, the `if` chain after the `select` might be a bit complicated, I guess.  I personally like to return fast for a special condition than to keep caring of `BloackingWait` through the function.

I can update the PR on your call.